books/workshops/references/xdoc-generation.lisp

```diff
@@ -0,0 +1,154 @@
+      (strip-brackets title-str)
+    "Untitled"))
+
+(defun get-field-value (field-name fields)
+  "Get the value of a field from the fields alist"
+  (let ((pair (assoc-equal field-name fields)))
+    (if pair
+        (cdr pair)
+      nil)))
+
+(defun get-entry-field (field-name entry)
+  "Get a top-level field from a BibTeX entry (type, key, or fields)"
+  (let ((pair (assoc-equal field-name entry)))
+    (if pair
+        (cdr pair)
+      nil)))
+
+(defun format-year (year-str)
+  "Format year for display"
+  (if year-str
+      year-str
+    "Unknown Year"))
+
+(defun format-pages (pages-str)
+  "Format page numbers for display"
+  (if pages-str
+      (concatenate 'string "pp. " pages-str)
+    ""))
+
+(defun format-entry-type (entry-type)
+  "Format entry type for display"
+  (cond ((equal entry-type "article") "Journal Article")
+        ((equal entry-type "inproceedings") "Conference Paper")
+        ((equal entry-type "incollection") "Book Chapter")
+        ((equal entry-type "techreport") "Technical Report")
+        (t (concatenate 'string
+                        (string-upcase (subseq entry-type 0 1))
+                        (subseq entry-type 1 nil)))))
+
+(defun generate-bibtex-entry-html (entry)
+  "Generate HTML list item for a single BibTeX entry"
+  (let* ((entry-type (get-entry-field "type" entry))
+         (fields (get-entry-field "fields" entry))
+         (author (get-field-value "author" fields))
+         (title (get-field-value "title" fields))
+         (year (get-field-value "year" fields))
+         (pages (get-field-value "pages" fields))
+         (publisher (get-field-value "publisher" fields))
+         (booktitle (get-field-value "booktitle" fields))
+         (month (get-field-value "month" fields))
+         (isbn (get-field-value "isbn" fields))
+         (doi (get-field-value "doi" fields)))
+
+    (concatenate 'string
+                 "<li><b>" (format-title title) "</b>, "
+                 (format-author-names author)
+                 (if booktitle
+                     (concatenate 'string ", in <i>" (strip-brackets booktitle) "</i>")
+                   "")
+                 (if publisher
+                     (concatenate 'string ", " (strip-brackets publisher))
+                   "")
+                 (if year
+                     (concatenate 'string ", " (strip-brackets year))
+                   "")
+                 (if month
+                     (concatenate 'string ", " (strip-brackets month))
+                   "")
+                 (if pages
+                     (concatenate 'string ", " (format-pages pages))
+                   "")
+                 (if isbn
+                     (concatenate 'string ", ISBN: " (strip-brackets isbn))
+                   "")
+                 (if doi
+                     (concatenate 'string ", DOI: " (strip-brackets doi))
+                   "")
+                 ". [" (format-entry-type entry-type) "]"
+                 "</li>")))
```

Comment:
Perhaps it would have been easier to use the XDOC constructors (see `:doc constructors`). But this works :smile:.